### PR TITLE
Fixed sonar pulse getting deployed on land

### LINKED ISSUE #6145: Sonar pulse can be used as an overpowered spyplane
It can instantly reveal any location on the map once every 30 seconds.

This should be restricted to water only (or more specifically a list of terrain types specified in the power), and the shroud reveal should be removed.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3460120-sonar-pulse-can-be-used-as-an-overpowered-spyplane?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #17371: SpawnActorPower: cannot filter for Terrain
An SpawnActorPower can spawn the given unit on spaces the unit given normally wouldnt able to go to. This can potentially cause Vehicles on Water, Buildings on Water, inside Cliffs, etc. There should be an way to limit this, like it would check if space is buildable then spawn the building there, or check if space is passable for unit's movetype and block if uneligible.

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/6145.
Closes #17371.

### CODE REVIEW

**Comment by pchote** [comment]
Unfortunately this leaks state through the shroud: the cursor tells the player whether a cell hidden under the shroud is water or something else. We will also need to limit this to cells that are either currently visible or perhaps visible + fog.

**Comment by Mailaender** [comment]
Added a shroud check.

**Comment by MlemandPurrs** [comment]
does Terrain= only accept a single type or a list of comma seperated terrain ? asking because, see:
https://github.com/OpenRA/OpenRA/issues/17371

**Comment by Mailaender** [comment]
A list of comma seperated terrain types.

**Comment by MustaphaTR** [comment]
Can we put these checks under base SupportPower instead of SpawnActorPower? i can see Terrain checks being used for other powers too.

**Comment by pchote** [comment]
Adding those kind of checks to `SupportPower` would not be a good idea. Some of our powers (e.g. `ProduceActorPower`, `GrantPrerequisiteChargeDrainPower`) do not have targeting cursors, so it does not make sense for them to expose terrain types.

This PR is only applying the checks in the client-side order targeting: this means that players can modify their client to allow them to target the power anywhere on the map without desyncing the game. The support power code does not expose a way for implementations to tell the manager that a targeting request is invalid and should be rejected - this is why the iron curtain and chronosphere misfires occur. We really should fix that before we try to proceed here.

**Comment by Orb370** [comment]
Maybe the cursor shouldn't change when it's on land? I think the name makes it pretty obvious it needs to be used in the water.

**Review by obrakmann — COMMENTED** [review]
Works as advertised and seems reasonable. :+1:

**Review by abcdefg30 — CHANGES_REQUESTED** [review]
We should also perform those checks in `Activate` before anything is done by the power. That will alleviate the cheating concerns.

**Comment by Mailaender** [comment]
Added the checks to activation as well.

File: OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs [hunk 1]
```diff
@@ -50,30 +62,113 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
 			base.Activate(self, order, manager);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: We should probably call this after the checks, shouldn't we?

File: OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs [hunk 2]
```diff
@@ -50,30 +62,113 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
...
+
+			if (!info.AllowUnderShroud && self.World.ShroudObscures(cell))
+				return;
 
-			if (info.Actor != null)
+			if (info.Terrain != null && !info.Terrain.Contains(self.World.Map.GetTerrainInfo(cell).Type))
+				return;
+
+			if (info.Actor == null)
+				return;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: That field has `FieldLoader.Require` (and an `ActorReference`). Can this even be null?

**Review by abcdefg30 — COMMENTED** [review]
A few comments then this should be good to go. (Bonus points if you managed to reduce duplication.)

File: OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs (hunk 1 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: Maybe.

File: OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs (hunk 2 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: No. You are right. I removed it.

**Comment by Mailaender** [comment]
Removed the redundancy. I thought there might have been performance implications, but there aren't any.

File: OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs [hunk 3]
```diff
@@ -47,33 +59,117 @@ public SpawnActorPower(Actor self, SpawnActorPowerInfo info)
...
+			if (!world.Map.Contains(cell))
+				return info.BlockedCursor;
+
+			if (!info.AllowUnderShroud && world.ShroudObscures(cell))
+				return info.BlockedCursor;
+
+			if (info.Terrain != null && !info.Terrain.Contains(world.Map.GetTerrainInfo(cell).Type))
+				return info.BlockedCursor;
+
+			return info.Cursor;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: We should be able to simplify this to
```suggestion
			return power.Validate(world, info, cell) ? info.Cursor : info.BlockedCursor;
```
now.

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: Yes,  you are correct.

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/SupportPowers/SpawnActorPower.cs]: Bonus points to you.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bca00c16b00803b3878673b1d8679af57be129bd...fd1b201ed400447bbbee0c5de65bad067b7ab01c)

**Pull request merged**